core/src/main/java/com/confighub/core/store/Store.java

```diff
@@ -4266,50 +4267,43 @@ else if ( max < 1 )
             max = 10;
         }
 
+        HashMap<String, Object> userParams = new HashMap<>(baseUserParams);
         StringBuilder hql = new StringBuilder();
         hql.append( baseHql );
 
         if ( 0 == starting && 0 == direction )
         {
-            hql.append( "ORDER BY id DESC" );
+            hql.append( " ORDER BY id DESC" );
         }
         else if ( direction > 0 )
         {
-            hql.append( "AND id < " ).append( starting ).append( " " );
-            hql.append( "ORDER BY id DESC" );
+            hql.append( " AND id < :startingId ORDER BY id DESC" );
+            userParams.put("startingId", starting);
         }
         else
         {
-            hql.append( "AND id > " ).append( starting ).append( " " );
-            hql.append( "ORDER BY id ASC" );
+            hql.append( " AND id > :startingId ORDER BY id ASC" );
+            userParams.put("startingId", starting);
         }
 
         Query query = em.createQuery( hql.toString(), RevisionEntry.class )
                         .setLockMode( LockModeType.NONE )
                         .setMaxResults( max );
+        userParams.forEach( ( param, value ) -> query.setParameter( param, value ) );
         List<RevisionEntry> revs = query.getResultList();
 
-        if ( direction < 0 && revs.size() < max )
+        if ( revs.size() < max && direction != 0  )
         {
+            userParams = new HashMap<>(baseUserParams);
             hql = new StringBuilder();
             hql.append( baseHql );
-            hql.append( "ORDER BY id DESC" );
+            hql.append( " ORDER BY id " + (direction > 0 ? "ASC" : "DESC") );
 
-            query = em.createQuery( hql.toString(), RevisionEntry.class )
-                      .setLockMode( LockModeType.NONE )
-                      .setMaxResults( max );
-            revs = query.getResultList();
-        }
-        else if ( direction > 0 && revs.size() < max )
-        {
-            hql = new StringBuilder();
-            hql.append( baseHql );
-            hql.append( "ORDER BY id ASC" );
-
-            query = em.createQuery( hql.toString(), RevisionEntry.class )
-                      .setLockMode( LockModeType.NONE )
-                      .setMaxResults( max );
-            revs = query.getResultList();
+            Query reQuery = em.createQuery( hql.toString(), RevisionEntry.class )
```

Comment:
Why is this a new variable now?

Reply:
The lambdas giveth, and the lambdas taketh away. Lambdas can only capture variables that are "final" or "effective final." Since the original query is mutated in the first lambda, it doesn't like it being used in a second lambda. Honestly, this'll probably be cleaner without lambdas.

Reply:
Which is a really stupid limitation. C++ and C# can both figure this out; get it together Java!